livekit-agents/livekit/agents/pipeline/room_io.py

```diff
@@ -0,0 +1,261 @@
+            self,
+            track_to_stream: Callable[
+                [rtc.RemoteTrack], rtc.AudioStream | rtc.VideoStream
+            ],
+            enabled: bool = False,
+        ) -> None:
+            self.track_to_stream = track_to_stream
+            self.enabled = enabled
+
+            self.track: rtc.RemoteTrack | None = None
+            self.stream_task: asyncio.Task | None = None
+            self._data_ch: aio.Chan[rtc.AudioFrame | rtc.VideoFrame] | None = None
+
+            if enabled:
+                self._data_ch = aio.Chan()
+
+        @property
+        def data_ch(self) -> aio.Chan | None:
+            return self._data_ch
+
+        def setup(self, track: rtc.RemoteTrack) -> None:
+            """Set up streaming for a new track"""
+            if track == self.track:
+                return
+
+            if self.stream_task is not None:
+                self.stream_task.cancel()
+
+            assert self._data_ch is not None
+            self.track = track
+            stream = self.track_to_stream(track)
+            self.stream_task = asyncio.create_task(self._stream_frames(stream))
+
+        async def _stream_frames(
+            self, stream: rtc.AudioStream | rtc.VideoStream
+        ) -> None:
+            assert self._data_ch is not None
+            async for event in stream:
+                self._data_ch.send_nowait(event.frame)
+
+    def __init__(
+        self,
+        room: rtc.Room,
+        participant_identity: Optional[str] = None,
+        *,
+        audio_enabled: bool = True,
+        video_enabled: bool = False,
+        audio_sample_rate: int = 16000,
+        audio_num_channels: int = 1,
+    ) -> None:
+        """
+        Args:
+            room: The LiveKit room to get streams from
+            participant_identity: Optional identity of the participant to get streams from.
+                                If None, will use the first participant that joins.
+            audio_enabled: Whether to enable audio input
+            video_enabled: Whether to enable video input
+            audio_sample_rate: Sample rate of the audio in Hz
+            audio_num_channels: Number of audio channels
+        """
+        self._room = room
+        self._expected_identity = participant_identity
+        self._participant: rtc.RemoteParticipant | None = None
+        self._closed = False
+
+        # set up track streamers
+        self._audio_streamer = self._RemoteTrackHandler(
+            track_to_stream=partial(
+                rtc.AudioStream,
+                sample_rate=audio_sample_rate,
+                num_channels=audio_num_channels,
+            ),
+            enabled=audio_enabled,
+        )
+        self._video_streamer = self._RemoteTrackHandler(
+            track_to_stream=rtc.VideoStream,
+            enabled=video_enabled,
+        )
+
+        self._participant_ready = asyncio.Event()
+        self._room.on("participant_connected", self._on_participant_connected)
```

Comment:
Let's handle room reconnection. If you get a "reconnected" event, you have to republish the local track.
Also we will have to "refind" the track we want to subscribe to